pkg/chunkenc/symbols.go

```diff
@@ -0,0 +1,394 @@
+}
+
+// symbolizerFromCheckpoint builds symbolizer from the bytes generated during a checkpoint.
+func symbolizerFromCheckpoint(b []byte) *symbolizer {
+	if len(b) == 0 {
+		return newSymbolizer()
+	}
+
+	db := decbuf{b: b}
+	s := symbolizer{
+		symbolsMap: map[string]uint32{},
+	}
+	numLabels := db.uvarint()
+
+	for i := 0; i < numLabels; i++ {
+		label := string(db.bytes(db.uvarint()))
+		s.labels = append(s.labels, label)
+		s.symbolsMap[label] = uint32(i)
+		s.size += len(label)
+	}
+
+	return &s
+}
+
+// symbolizerFromEnc builds symbolizer from the bytes generated during serialization.
+func symbolizerFromEnc(b []byte, pool ReaderPool) (*symbolizer, error) {
+	db := decbuf{b: b}
+	numSymbols := db.uvarint()
+
+	b = db.b
+
+	reader, err := pool.GetReader(bytes.NewBuffer(b))
+	if err != nil {
+		return nil, err
+	}
+	defer pool.PutReader(reader)
+
+	var (
+		readBuf      [10]byte // Enough bytes to store one varint.
+		readBufValid int      // How many bytes are left in readBuf from previous read.
+		s            symbolizer
+		buf          []byte
+	)
+
+	s.compressedSize = len(b)
+
+	for i := 0; i < numSymbols; i++ {
+		var lWidth, labelSize, lastAttempt int
+		for lWidth == 0 { // Read until both varints have enough bytes.
+			n, err := reader.Read(readBuf[readBufValid:])
+			readBufValid += n
+			if err != nil {
+				if err != io.EOF {
+					return nil, err
+				}
+				if readBufValid == 0 { // Got EOF and no data in the buffer.
+					return nil, fmt.Errorf("got unexpected EOF")
+				}
+				if readBufValid == lastAttempt { // Got EOF and could not parse same data last time.
+					return nil, fmt.Errorf("invalid non-indexed labels block in chunk")
+				}
+			}
+			var l uint64
+			l, lWidth = binary.Uvarint(readBuf[:readBufValid])
+			labelSize = int(l)
+			lastAttempt = readBufValid
+		}
+
+		// If the buffer is not yet initialize or too small, we get a new one.
+		if buf == nil || labelSize > cap(buf) {
+			// in case of a replacement we replace back the buffer in the pool
+			if buf != nil {
+				BytesBufferPool.Put(buf)
+			}
+			buf = BytesBufferPool.Get(labelSize).([]byte)
+			if labelSize > cap(buf) {
+				return nil, fmt.Errorf("could not get a label buffer of size %d, actual %d", labelSize, cap(buf))
+			}
+		}
+		buf = buf[:labelSize]
+		// Take however many bytes are left in the read buffer.
+		n := copy(buf, readBuf[lWidth:readBufValid])
+		// Shift down what is still left in the fixed-size read buffer, if any.
+		readBufValid = copy(readBuf[:], readBuf[lWidth+n:readBufValid])
+
+		// Then process reading the line.
+		for n < labelSize {
+			r, err := reader.Read(buf[n:labelSize])
+			n += r
+			if err != nil {
+				// We might get EOF after reading enough bytes to fill the buffer, which is OK.
+				// EOF and zero bytes read when the buffer isn't full is an error.
+				if err == io.EOF && r != 0 {
+					continue
+				}
+				return nil, err
+			}
+		}
+		s.labels = append(s.labels, string(buf))
```

Comment:
should we also populate `symbolsMap` to be consistent?

Reply:
We load data from encoded bytes from a flushed chunk, so we only need the `labels` list for queries, so I am not populating `symbolsMap` here to save some space and processing.